src/mcpwm.rs

```diff
@@ -0,0 +1,802 @@
+            }
+
+            if config.lowest_frequency > mcpwm_module.borrow().operator_source_frequency.Hz() {
+                // Can not specify a lowest_frequency larger than the corresponding value for
+                // the parent MCPWM module. Use `Mcpwm::lowest_frequency` to enable higher frequencies
+                return Err(EspError::from(ESP_ERR_INVALID_ARG).unwrap());
+            }
+
+            let resolution = u32::from(config.lowest_frequency) * MAX_PWM_TIMER_PERIOD;
+            unsafe {
+                esp_idf_sys::mcpwm_timer_set_resolution(U::unit(), O::timer(), resolution);
+            }
+        }
+
+        // TODO: Handle/document half pwm frequency when counter_mode = UpDown?
+
+        esp!(unsafe {
+            esp_idf_sys::mcpwm_init(
+                U::unit(),
+                O::timer(),
+                &mcpwm_config_t {
+                    frequency: config.frequency.into(),
+                    cmpr_a: config.duty_a,
+                    cmpr_b: config.duty_b,
+                    duty_mode: config.duty_mode.into(),
+                    counter_mode: config.counter_mode.into(),
+                },
+            )
+        })?;
+
+        match config.deadtime {
+            None => unsafe {
+                // Only way this can faild is if an invalid timer or unit is specified
+                // which we know can not happen. So we don't have to check for errors
+                mcpwm_deadtime_disable(U::unit(), O::timer());
+            },
+            Some(config) => {
+                let DeadtimeArgs { red, fed, mode } = config.as_args();
+                unsafe {
+                    mcpwm_deadtime_enable(U::unit(), O::timer(), mode, red.into(), fed.into());
+                }
+            }
+        }
+
+        let pin_a: Option<PA> = pin_a.into();
+        let pin_b: Option<PB> = pin_b.into();
+
+        if let Some(pin_a) = &pin_a {
+            let io_signal = O::signal_a();
+            esp!(unsafe { esp_idf_sys::mcpwm_gpio_init(U::unit(), io_signal, pin_a.pin()) })?;
+        }
+
+        if let Some(pin_b) = &pin_b {
+            let io_signal = O::signal_b();
+            esp!(unsafe { esp_idf_sys::mcpwm_gpio_init(U::unit(), io_signal, pin_b.pin()) })?;
+        }
+
+        Ok(Self {
+            _instance: operator,
+            _unit: U::default(),
+            _mcpwm_module: mcpwm_module,
+            _pin_a: pin_a,
+            _pin_b: pin_b,
+        })
+    }
+
+    pub fn release(self) -> (O, Option<PA>, Option<PB>) {
+        // mcpwm_stop will only fail when invalid args are given
+        esp!(unsafe { mcpwm_stop(U::unit(), O::timer()) }).unwrap();
+
+        // TODO: Test and verify if this is the right way
+        if self._pin_a.is_some() {
+            // TODO: How to unset pin?
+            // let io_signal = O::signal_a();
+            // mcpwm_gpio_init(U::unit(), io_signal, -1)
+            // does not seem to be it...
+            todo!();
```

Comment:
It it at all possible to unset a pin?

If not, what will happen the first time when no pins are specified, will some pin be defaulted to? Do we have to force the user to specify two pins in order to overwrite the config?

Reply:
[This](https://github.com/espressif/esp-idf/blob/16a4ee7c36a848ca155791677ce011f3ca75c519/components/driver/mcpwm.c#L140-L144) is how `mcpwm_gpio_init` is currently implemented in master

Reply:
The [Arduino libraries](https://github.com/wemos/Arduino_ESP32/blob/master/cores/esp32/esp32-hal-matrix.c#L35) seems to set 0x100 (256) as out signal to detach pin if I understand things right

Reply:
Looking at i2c: 

https://github.com/esp-rs/esp-idf-hal/blob/master/src/i2c.rs#L222-L228

there seems to be a commented call to `pin.reset()` however from what I can tell that seems to be implemented on the concrete type and not available generically...

Reply:
> there seems to be a commented call to pin.reset() however from what I can tell that seems to be implemented on the concrete type and not available generically...

Ah, this seems like an oversight. It should be possible to move that implementation to the `Pin` trait. I think you'll also need to move `disable_interrupt` there too.

Reply:
Not sure about the reason behind it but it seems there was a `reset` method on `Pin` until 2d0554d